pkg/assets/crd.go

```diff
@@ -44,57 +42,82 @@ var (
 	}
 )
 
-func getCRD(clientset *apiext_clientset.Clientset, obj apiruntime.Object) error {
+func isEstablished(clientset *apiext_clientset.Clientset, obj apiruntime.Object) (bool, error) {
 	gv := obj.GetObjectKind().GroupVersionKind().GroupVersion()
 	switch gv.String() {
 	case "apiextensions.k8s.io/v1":
 		v1Obj := obj.(*apiextv1.CustomResourceDefinition)
-		_, err := clientset.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), v1Obj.Name, metav1.GetOptions{})
+		crd, err := clientset.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), v1Obj.Name, metav1.GetOptions{})
 		if err != nil {
 			lastErr := fmt.Errorf("error getting CustomResourceDefinition %s: %v", v1Obj.Name, err)
 			logrus.Infof("getting openshift CRD status %v", lastErr)
-			return lastErr
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, lastErr
+		}
+		for _, condition := range crd.Status.Conditions {
+			if condition.Type == apiextv1.Established && condition.Status == apiextv1.ConditionTrue {
+				return true, nil
+			}
 		}
 	case "apiextensions.k8s.io/v1beta1":
 		v1beta1Obj := obj.(*apiextv1beta1.CustomResourceDefinition)
-		_, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), v1beta1Obj.Name, metav1.GetOptions{})
+		crd, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), v1beta1Obj.Name, metav1.GetOptions{})
 		if err != nil {
 			lastErr := fmt.Errorf("error getting CustomResourceDefinition %s: %v", v1beta1Obj.Name, err)
 			logrus.Infof("getting openshift CRD status %v", lastErr)
-			return lastErr
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, lastErr
+		}
+		for _, condition := range crd.Status.Conditions {
+			if condition.Type == apiextv1beta1.Established && condition.Status == apiextv1beta1.ConditionTrue {
+				return true, nil
+			}
 		}
 	default:
 		// panic("unknown type %s", t)
 	}
-	return nil
+	return false, nil
 }
 
-func waitForCRD(client apiextclientv1beta1.CustomResourceDefinitionsGetter, resource *apiextv1beta1.CustomResourceDefinition) error {
-	var lastErr error
-	if err := wait.Poll(customResourceReadyInterval, customResourceReadyTimeout, func() (bool, error) {
-		crd, err := client.CustomResourceDefinitions().Get(context.TODO(), resource.Name, metav1.GetOptions{})
+func WaitForCrdsEstablished(cfg *config.MicroshiftConfig) error {
+	lock.Lock()
```

Comment:
what is the lock for?

Reply:
Unneeded I think, will drop.  Looks like the lock exists to flow-control calls to the api server for `readerApplier`'s.  The crd manager doesn't implement that interface though, so maybe it's not needed at all?